runtime/tests/interpreter/memory_metering_test.go

```diff
@@ -93,10 +93,7 @@ func TestRuntimeArrayMetering(t *testing.T) {
 		_, err := inter.Invoke("main")
 		require.NoError(t, err)
 
-		// TODO:
-		// Iteration create new values for array typed elements.
-		// Currently, these are not metered. Only the initial 4-values are captured.
-		assert.Equal(t, uint64(4), meter.getMemory(common.MemoryKindArray))
+		assert.Equal(t, uint64(16), meter.getMemory(common.MemoryKindArray))
```

Comment:
Oh wow, that increased the values drastically. I was expecting it to be 10 (4-initial values, 3 for iteration, and 3 for assignment). Looking at the above test, it seems the allocations on array-literal also has increased.

Maybe good to add comment with a break-down on where the allocations happened?

Reply:
Thanks @dsainati1 for the complex Analysis. This is very useful!

Since it's the interpreter-implementation that allocates more memory than what's probably needed, and the memory metering correctly captures them, I would think it's close to actual memory usage, and should be fine. (Given our goal is to be as close as possible to actual usage.

But definitely we should optimize these, maybe as a follow-up task after secure cadence. Couple items I could think of inspired from these findings are:
- DynamicType checking - This one have been under radar for a while. I think https://github.com/onflow/cadence/issues/1217 (this guy again!) would actually solve this. See the `CheckValueTransferTargetType` alloc improvements on the benchmark results.
- Perhaps we don't need to transfer/copy for literals. Can skip, as an optimization.